ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1881,18 +2057,318 @@ void shutdown() throws InterruptedException {
     } finally {
       lock.unlock();
     }
+    shouldTerminateCurrentCycle = true;
     if (networkClient != null) {
       networkClient.close();
     }
     shutdownLatch.await();
   }
 
+  class RemoteReplicaGroupPoller {
+    private final int maxIterationsPerGroup;
+    private boolean allReplicasCaughtUpEarly;
+    private final Map<Integer, List<RemoteReplicaInfo>> groupIdToRemoteReplicaMap;
+    private final Map<DataNodeId, Integer> standByGroupIds;
+    private final Map<Integer, RemoteReplicaGroup> inflightGroups;
+    private final Map<Integer, Integer> groupIdIterationCountMap;
+    private final Map<DataNodeId, Set<RemoteReplicaInfo>> standByReplicaQueue;
+    private final Map<RemoteReplicaInfo, Long> replicaToThrottledTill;
+
+    RemoteReplicaGroupPoller( ) {
+      maxIterationsPerGroup = maxIterationsPerGroupPerCycle;
+      allReplicasCaughtUpEarly = false;
+      groupIdToRemoteReplicaMap = new HashMap<>();
+      standByGroupIds = new HashMap<>();
+      inflightGroups = new HashMap<>();
+      groupIdIterationCountMap = new HashMap<>();
+      standByReplicaQueue = new HashMap<>();
+      replicaToThrottledTill = new HashMap<>();
+      createGroupIds();
+    }
+
+    boolean allReplicasCaughtUpEarly() {
+      return allReplicasCaughtUpEarly;
+    }
+
+    Map<Integer, List<RemoteReplicaInfo>> getGroupIdToRemoteReplicaMap() {
+      return groupIdToRemoteReplicaMap;
+    }
+
+    Map<DataNodeId, Integer> getStandByGroupIds() {
+      return standByGroupIds;
+    }
+
+    Map<Integer, Integer> getGroupIdIterationCountMap() {
+      return groupIdIterationCountMap;
+    }
+
+    Map<DataNodeId, Set<RemoteReplicaInfo>> getStandByReplicaQueue() {
+      return standByReplicaQueue;
+    }
+
+    Map<RemoteReplicaInfo, Long> getReplicaToThrottledTill() {
+      return replicaToThrottledTill;
+    }
+
+    int allProcessedReplicaCount(){
+      return groupIdToRemoteReplicaMap.values().stream().mapToInt(List::size).sum();
+    }
+
+    /**
+     * Generates group ids for active iteration for replicas, all replicas are mapped to one group id only
+     * Generates group ids for all datanodes for standby no progress groups
+     * All group ids are unique
+     * All replicas are mapped to particular group id and each datanode has one stand by remote replica group id
+     * so we make less number of parallel calls to remote hosts
+     * {@link #groupIdToRemoteReplicaMap} this map will be populated with group ids to replicas map
+     * {@link #standByGroupIds}  this map will be populated with datanode to standby group id
+     */
+    private void createGroupIds() {
+      Map<DataNodeId, List<RemoteReplicaInfo>> dataNodeToRemoteReplicaInfo = selectReplicas(getRemoteReplicaInfos());
+
```

Comment:
groupIdToRemoteReplicaMap : id -> replicas
standByGroupIds : node -> id
states: inflight
var: num_iter, throttled_till

How about this ? Lmk your thoughts

List<PeerReplicas> listPeerReplicas; // iterate on this list in the do-while loop

class RemoteReplicaGroupInfo {
 RemoteReplicaGroup group;
 int numIterations;
 long throttleuntil;
 // use the group.state to check if inflight, remove the inflight state
}

class PeerReplicas {
  DataNodeId datanode;
  List<RemoteReplicaInfo> remoteReplicas;
  List<RemoteReplicaGroupInfo> activeGroups; // iterate on this for each peer
  List<RemoteReplicaGroupInfo> standByGroups; // iterate on this for each peer
}